### DISCUSSION TITLE
Enable datasets with few entries for live system integration

### DISCUSSION THREAD

**Discussion by Corinna-Frank** [neuralhydrology/neuralhydrology#132] [Ideas]
Hi,

I am testing to use the package in a live-system setup such that the model is trained in advance and one prediction calculated every day. For this setup the dataset is updated every day and includes only the newest data, i.e., only a single timestep. This creates an error as the frequency of the data is inferred from the dataset itself using pandas, and pandas throws an error if the time-series is less than three elements long:

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> 
> File \evaluation.py:135, in load_dataset(cfg, run_dir)
>     132 def load_dataset (cfg, run_dir):
>     134     scaler = load_scaler(Path(run_dir))
> --> 135     dataset = get_dataset(cfg, is_train=False, period='test', scaler=scaler)
>     137     return dataset
> 
> File \neuralhydrology\datasetzoo\__init__.py:83, in get_dataset(cfg, is_train, period, basin, additional_features, id_to_int, scaler)
>      80 else:
>      81     raise NotImplementedError(f"No dataset class implemented for dataset {cfg.dataset}")
> ---> 83 ds = Dataset(cfg=cfg,
>      84              is_train=is_train,
>      85              period=period,
>      86              basin=basin,
>      87              additional_features=additional_features,
>      88              id_to_int=id_to_int,
>      89              scaler=scaler)
>      90 return ds
> 
> File \neuralhydrology\datasetzoo\genericdataset.py:61, in GenericDataset.__init__(self, cfg, is_train, period, basin, additional_features, id_to_int, scaler)
>      53 def __init__(self,
>      54              cfg: Config,
>      55              is_train: bool,
>    (...)
>      59              id_to_int: Dict[str, int] = {},
>      60              scaler: Dict[str, Union[pd.Series, xarray.DataArray]] = {}):
> ---> 61     super(GenericDataset, self).__init__(cfg=cfg,
>      62                                          is_train=is_train,
>      63                                          period=period,
>      64                                          basin=basin,
>      65                                          additional_features=additional_features,
>      66                                          id_to_int=id_to_int,
>      67                                          scaler=scaler)
> 
> File \neuralhydrology\datasetzoo\basedataset.py:131, in BaseDataset.__init__(self, cfg, is_train, period, basin, additional_features, id_to_int, scaler)
>     128         self._create_id_to_int()
>     130 # load and preprocess data
> --> 131 self._load_data()
>     133 if self.is_train:
>     134     self._dump_scaler()
> 
> File \neuralhydrology\datasetzoo\basedataset.py:650, in BaseDataset._load_data(self)
>     646 def _load_data(self):
>     647     # load attributes first to sanity-check those features before doing the compute expensive time series loading
>     648     self._load_combined_attributes()
> --> 650     xr = self._load_or_create_xarray_dataset()
>     652     if self.cfg.loss.lower() in ['nse', 'weightednse']:
>     653         # get the std of the discharge for each basin, which is needed for the (weighted) NSE loss.
>     654         self._calculate_per_basin_std(xr)
> 
> File \neuralhydrology\datasetzoo\basedataset.py:326, in BaseDataset._load_or_create_xarray_dataset(self)
>     321 start_dates = self.start_and_end_dates[basin]["start_dates"]
>     322 end_dates = [
>     323     date + pd.Timedelta(days=1, seconds=-1) for date in self.start_and_end_dates[basin]["end_dates"]
>     324 ]
> --> 326 native_frequency = utils.infer_frequency(df.index)
>     327 if not self.frequencies:
>     328     self.frequencies = [native_frequency]  # use df's native resolution by default
> 
> File \neuralhydrology\datautils\utils.py:170, in infer_frequency(index)
>     151 def infer_frequency(index: Union[pd.DatetimeIndex, np.ndarray]) -> str:
>     152     """Infer the frequency of an index of a pandas DataFrame/Series or xarray DataArray.
>     153 
>     154     Parameters
>    (...)
>     168         If the frequency cannot be inferred from the index or is zero.
>     169     """
> --> 170     native_frequency = pd.infer_freq(index)
>     171     if native_frequency is None:
>     172         raise ValueError(f'Cannot infer a legal frequency from dataset: {native_frequency}.')
> 
> File \pandas\tseries\frequencies.py:201, in infer_freq(index, warn)
>     198 if not isinstance(index, DatetimeIndex):
>     199     index = DatetimeIndex(index)
> --> 201 inferer = _FrequencyInferer(index, warn=warn)
>     202 return inferer.get_freq()
> 
> File \pandas\tseries\frequencies.py:242, in _FrequencyInferer.__init__(self, index, warn)
>     239 self.warn = warn
>     241 if len(index) < 3:
> --> 242     raise ValueError("Need at least 3 dates to infer frequency")
>     244 self.is_monotonic = (
>     245     self.index._is_monotonic_increasing or self.index._is_monotonic_decreasing
>     246 )
> 
> ValueError: Need at least 3 dates to infer frequency

Is it possible to pass the frequency manually via the config when it cannot be inferred? Looking at the BaseDataset in \datasetzoo\basedataset.py:326 it seems like the native frequency is always inferred. It might be interesting to add another keyword in the config file, such as "overwrite_native_frequency" (should be False by default) for which the native frequency will not be calculated. The basedataset could for instance be adapted as follows:

```python
if not self.frequencies:
        native_frequency = utils.infer_frequency(df.index)
        self.frequencies = [native_frequency]
else: # frequency was passed in config
        if self.cfg.overwrite_native_frequency:
                native_frequency = self.frequencies[0]
        else:
                native_frequency = utils.infer_frequency(df.index)

```

Best regards,
Corinna

**Comment by gauchm** [comment]
I am not sure if I'm missing something in your setup, but IIUC, you are creating a new individual dataset for each day, which sounds cumbersome to me already from a data storage perspective. 
Couldn't you just keep appending the incoming days to one dataset and only change the `test_start_date`/`test_end_date` in the config to always just include the current day?

**Reply by Corinna-Frank** [reply]
Sure, this is definitely possible, thanks for the advice. The initial thought was to only store the bare minimum of data and avoid having a file that increases in size over time. But since the netcdf files are small anyways it might not create a problem for the system.

**Reply by gauchm** [reply]
You could of course store the dataset in a first-in-first-out fashion that only keeps the most recent 3 (or more) days if storage size is a concern :) Maybe this is even helpful for debugging in case you want to go back a few days and reproduce a past prediction.
